File: src_fortran/sub/fixunt.f
c
c
c
c
c
c
      subroutine FIXUNT(b,neqs,nshot,nl,ksta,scale,
     &                  vdamp,itotmodels,inltot,nlfirst)
c
c---- restore solution to proper units
c
      implicit none
      integer neqs,nshot,nl,ksta
      real b(4*neqs+nshot+nl+ksta),scale(7)
      integer nlfirst,itotmodels,inltot
      real vdamp(itotmodels,inltot)
      integer i,j,k,l,m
      i=0
      if(neqs.le.0) goto 10
      do 1 j=1,neqs
      do 1 k=1,4
      i=i+1
    1 b(i)=b(i)*scale(k)
10    continue
      if(nshot.le.0) goto 3
      do 2 j=1,nshot
      i=i+1
    2 b(i)=b(i)*scale(7)
3     if(scale(6).eq.0.) return
      l=0
      m=1
      do 5 j=1,nl
      i=i+1
      l=l+1
      if(l.gt.nlfirst) then
         m=m+1
	 l=1
      endif
chrm      write(6,*)m,'  ',l
5     b(i)=b(i)*scale(6)/vdamp(m,l)
      if(ksta.eq.0) return
      do 4 j=1,ksta
      i=i+1
    4 b(i)=b(i)*scale(5)
      return
      end ! of subr. fixunt

File: src/sub/fixunt.c
#include <stddef.h>

void fixunt(double *b, int neqs, int nshot, int nl, int ksta,
            const double *scale, const double *vdamp, int itotmodels,
            int inltot, int nlfirst) {
    int i, j, k, l, m;

    if (b == NULL || scale == NULL || vdamp == NULL) {
        return;
    }

    i = -1;

    if (neqs > 0) {
        for (j = 0; j < neqs; ++j) {
            for (k = 0; k < 4; ++k) {
                ++i;
                b[i] *= scale[k];
            }
        }
    }

    if (nshot > 0) {
        for (j = 0; j < nshot; ++j) {
            ++i;
            b[i] *= scale[6];
        }
    }

    if (scale[5] == 0.0f) {
        return;
    }

    l = 0;
    m = 1;
    for (j = 0; j < nl; ++j) {
        ++i;
        ++l;
        if (l > nlfirst) {
            ++m;
            l = 1;
        }
        if (m < 1 || m > itotmodels || l < 1 || l > inltot) {
            return;
        }
        double vd = vdamp[(m - 1) * inltot + (l - 1)];
        if (vd != 0.0f) {
            b[i] *= scale[5] / vd;
        }
    }

    if (ksta > 0) {
        for (j = 0; j < ksta; ++j) {
            ++i;
            b[i] *= scale[4];
        }
    }
}
